src/Share.php

```diff
@@ -0,0 +1,161 @@
+     */
+    public function __construct(
+        ApiPermission $apiPermission,
+        string        $resourceId,
+        string        $driveId,
+        array         $connectionConfig,
+        string        $serviceUrl,
+        string        &$accessToken
+    ) {
+        $this->apiPermission = $apiPermission;
+        $this->driveId = $driveId;
+        if (!is_string($apiPermission->getId())) {
+            throw new InvalidResponseException(
+                "Invalid id returned for permission '" . print_r($apiPermission->getId(), true) . "'"
+            );
+        }
+
+        $this->resourceId = $resourceId;
+        $this->accessToken = &$accessToken;
+        $this->serviceUrl = $serviceUrl;
+        if (!Ocis::isConnectionConfigValid($connectionConfig)) {
+            throw new \InvalidArgumentException('connection configuration not valid');
+        }
+        $this->graphApiConfig = Configuration::getDefaultConfiguration()
+            ->setHost($this->serviceUrl . '/graph');
+
+        $this->connectionConfig = $connectionConfig;
+    }
+
+    protected function getDrivesPermissionsApi(): DrivesPermissionsApi
+    {
+        $guzzle = new Client(
+            Ocis::createGuzzleConfig($this->connectionConfig, $this->accessToken)
+        );
+        if (array_key_exists('drivesPermissionsApi', $this->connectionConfig)) {
+            return $this->connectionConfig['drivesPermissionsApi'];
+        } else {
+            return new DrivesPermissionsApi(
+                $guzzle,
+                $this->graphApiConfig
+            );
+        }
+    }
+
+    public function getPermissionId(): string
+    {
+        // in the constructor the value is checked for being the right type, but phan does not know
+        // so simply cast to string
+        return (string)$this->apiPermission->getId();
+    }
+
+    public function getExpiry(): ?\DateTime
+    {
+        return $this->apiPermission->getExpirationDateTime();
+    }
+
+    /**
+     * permanently delete the current share or share link
+     *
+     * @throws BadRequestException
+     * @throws ForbiddenException
+     * @throws HttpException
+     * @throws NotFoundException
+     * @throws UnauthorizedException
+     * @throws InvalidResponseException
+     */
+    public function delete(): bool
+    {
+        try {
+            $this->getDrivesPermissionsApi()->deletePermission(
+                $this->driveId,
+                $this->resourceId,
+                $this->getPermissionId()
+            );
+        } catch (ApiException $e) {
+            throw ExceptionHelper::getHttpErrorException($e);
+        }
+        return true;
+    }
+
+    /**
+     * @throws UnauthorizedException
+     * @throws ForbiddenException
+     * @throws InvalidResponseException
+     * @throws BadRequestException
+     * @throws HttpException
+     * @throws NotFoundException
+     */
+    public function setExpiration(\DateTime $expiration): bool
```

Comment:
we can set role and expiration for share and sharelink, why setexpiration and setrole in different class?

Reply:
I've improved the comments of some functions, hope that explains it